### PR TITLE
Fixes #8: Show logs in UI

### LINKED ISSUE #8: Put link in batch job
We would like the link of the batch job we kicked off through Airflow to be shown in the UI. This will need to post to XCom. @vdat-coatue will help.

### PR DESCRIPTION
**Pull request by gmarks-coatue**
For the last run of each DAG, check every task matching `log_*` for the XCom key `log_urls` (the name of this key as well as the prefix to match tasks by are configurable as `xcom_log_urls_key` and `log_task_prefix`, respectively). For each matching task which has the given key, we interpret the corresponding value as a list of pairs of (name, URL) and link to the URL displaying the given name. Ex. If I want to link to the batch job of a DAG and the Cloudwatch log page, I would create a task, `log_stuff`, and have it  `xcom_push(key='log_urls', value=[('Batch Job', <batch-url>), ('Cloudwatch Log', <cloudwatch-url>)])`. When this task runs, the corresponding logs would appear in the last run column.

Before:
Screen Shot 2019-07-30 at 11 14 49 AM

After:
Screen Shot 2019-07-30 at 4 33 28 PM
In the above image, the second DAG in the picture has two tasks prepended with `log_`: `log_first` and `log_second`. As can be seen, these two are listed separately and their `task_id`s are included. The fourth DAG in the picture has only one task prepended with `log_`, so its `task_id` is not shown.

### CODE REVIEW

File: airflow/config_templates/default_airflow.cfg [hunk 1]
```diff
@@ -182,6 +182,11 @@ worker_precheck = True
 # When discovering DAGs, ignore any files that don't contain the strings `DAG` and `airflow`.
 dag_discovery_safe_mode = True
 
+# Prefix for tasks with link to log in XCom.
```
Comment (pathikrit) [inline airflow/config_templates/default_airflow.cfg]: the code comment is not that clear

File: airflow/www/views.py [hunk 2]
```diff
@@ -2200,8 +2201,22 @@ def get_int_arg(value, default=0):
...
+            xcom_log_urls_key=conf.get('core', 'xcom_log_urls_key'),
+            get_xcom=functools.partial(get_xcom, session),
             auto_complete_data=auto_complete_data)
 
+def get_xcom(session, dag_id, task_id, dttm):
+    XCom = models.XCom
+    xcomlist = session.query(XCom).filter(
+        XCom.dag_id == dag_id, XCom.task_id == task_id,
+        XCom.execution_date == dttm).all()
+    attributes = {}
```
Comment (pathikrit) [inline airflow/www/views.py]: do set comprehension:
```
return {xcom.key: xcom.value for xcom in xcomlist if not xcom.key.startswith('_')}
```

File: airflow/www/views.py [hunk 3]
```diff
@@ -2200,8 +2201,22 @@ def get_int_arg(value, default=0):
...
+    for xcom in xcomlist:
+        if not xcom.key.startswith('_'):
+            attributes[xcom.key] = xcom.value
```
Comment (pathikrit) [inline airflow/www/views.py]: code comment about this startswith

Comment (gmarks-coatue) [reply airflow/www/views.py]: i think it shouldn't be there, I copied this from the `XCom` log page for a given task and I think keys starting with underscore are meant for inter-task use or something.

File: airflow/www/views.py [hunk 4]
```diff
@@ -2200,8 +2201,18 @@ def get_int_arg(value, default=0):
             stringify_timedelta=stringify_timedelta,
             get_human_readable_cron=cron.get_human_readable_cron,
             dagbag=dagbag,
+            log_prefix=conf.get('core', 'log_task_prefix'),
+            xcom_log_urls_key=conf.get('core', 'xcom_log_urls_key'),
+            get_xcom=functools.partial(get_xcom, session),
             auto_complete_data=auto_complete_data)
 
+def get_xcom(session, dag_id, task_id, dttm):
```
Comment (pathikrit) [inline airflow/www/views.py]: code comment what this does

File: airflow/example_dags/example_log_onetask.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+# -*- coding: utf-8 -*-
+#
```
Comment (pathikrit) [inline airflow/example_dags/example_log_onetask.py]: Can you show @awong-coatue , @vdat-coatue how this works? Like what do we have to do to get the logs to show up?

Comment (gmarks-coatue) [reply airflow/example_dags/example_log_onetask.py]: I've explained above and also to @vdat-coatue, and I'm also in the process of writing a PR to add the functionality to `AWSBatchOperator`.

Comment (gmarks-coatue) [reply airflow/example_dags/example_log_onetask.py]: @pathikrit see https://github.com/coatue/datasyncs/pull/565, the corresponding PR for `AWSBatchOperator` to work nicely with this.

**Pull request merged**